File: Locality-Sensitive-Hashing.py
from pyspark import SparkContext
import time,sys
from itertools import combinations

inputfile = sys.argv[1]
outputfile = sys.argv[2]

# inputfile = "/Users/peiwendu/Downloads/public_data/yelp_train.csv"
# outputfile = "peiwen_du_task1"


sc = SparkContext(appName="inf553_hw3",master="local[*]")
sc.setLogLevel("WARN")
list()

# def minHash(map):
#     user_of_business = list(map)
#     # row
#     m = len(users)
#     # hash_function
#     hashes = [[913, 901, 24593], [14, 23, 769], [1, 101, 193], [17, 91, 1543],
#               [387, 552, 98317], [11, 37, 3079], [2, 63, 97], [41, 67, 6151],
#               [91, 29, 12289], [3, 79, 53], [73, 803, 49157], [8, 119, 389]]
#     signature = dict()
#     for j,p in enumerate(hashes):
#         signature[j] = m+1
#         for i in user_of_business:
#             signature[j] = min(signature[j],(p[0]*i+p[1])%p[2]%m)
#         # print(signature[j])
#     return [signature[key] for key in signature]

# def findCandidate(map,hashes,m):
#     bid1=map[0][0]
#     bid2=map[1][0]
#     # issimilar = False
#     similarity = 0
#     if bid1!=bid2:
#         maxlen = max(len(list(map[0][1])), len(list(map[1][1])))
#         minlen = max(len(list(map[0][1])), len(list(map[1][1])))
#         if minlen >= maxlen * 0.5:
#             # character1 = minHash(list(map[0][1]))
#             # character2 = minHash(list(map[1][1]))
#             character1 = [hashfunction(list(map[0][1]),has,m) for has in hashes]
#             character2 = [hashfunction(list(map[1][1]),has,m) for has in hashes]
#             for i in range(0,12,2):
#                 if character1[i:i+2] == character2[i:i+2]:
#                     similarity = len(set(character1).intersection(set(character2))) / len(set(character1).union(set(character2)))
#                     break
#     if similarity>=0.5:
#         print(similarity)
#         # return tuple(sorted((bid1,bid2)))
#         return (tuple(sorted([bid1,bid2])),similarity)
#     else:
#         return None

def hashfunction(x,has,m):
    return min([((has[0] * u + has[1]) % has[2]) % m for u in x])

def hashtosamebin(map):
    bid = map[0]
    signatiure = map[1]
    bins = []
    for i in range(0,20,2):
        bins.append(((i/2,tuple(signatiure[i:i+2])),[bid]))
    return bins

def findSimilarity(map,b_c,buineses):
    bid1 = map[0]
    bid2 = map[1]
    character1 = set(b_c[bid1])
    character2 = set(b_c[bid2])
    # print(len(character1.intersection(character2)))
    # print(len(character1.union(character2)))
    similarity = len(character1.intersection(character2))/len(character1.union(character2))
    # print(similarity)
    return (sorted([buineses[bid1],buineses[bid2]]),similarity)


start = time.time()
test_data = sc.textFile(inputfile).filter(lambda x:"user_id, business_id, stars" not in x).map(lambda x:x.split(","))
users = test_data.map(lambda x:x[0]).distinct().collect()
businesses = test_data.map(lambda x:x[1]).distinct().collect()
# print(len(businesses))
business_index = dict()
user_index = dict()
for i,bid in enumerate(businesses):
    business_index[bid] = i
for i,uid in enumerate(users):
    user_index[uid] = i
character_matrix = test_data.map(lambda x:(business_index[x[1]],[user_index[x[0]]])).reduceByKey(lambda x, y: x+y)
characters = character_matrix.collect()
b_c = dict()
for b in characters:
    b_c[b[0]] = b[1]

m= len(users)

hashes = [[913, 51, 24593], [54, 23, 769], [622,321,6291469], [1, 101, 193], [17, 91, 1543],[765,345,3145739],
              [387, 552, 98317], [11, 37, 3079], [2, 63, 97], [41, 67, 6151], [543,342,1572869],
              [291, 29, 12289], [3, 79, 53], [473, 83, 49157], [8, 119, 389], [321,445,196613], [951,340,393241],[531,300,786433],[224,21,25165843],[987,21,12582917]]


signatiure = character_matrix.map(lambda x:(x[0],[hashfunction(x[1],has,m) for has in hashes]))

candidate_pair = signatiure.flatMap(hashtosamebin).reduceByKey(lambda x,y: x+y).filter(lambda x:len(x[1]) >1).flatMap(lambda x:list(combinations(x[1],2))).map(lambda x:tuple(sorted(x))).distinct()

similar_pair = candidate_pair.map(lambda x:findSimilarity(x,b_c,businesses)).filter(lambda x:x[1]>=0.5).sortBy(lambda x:x[0][0])
# similar_pair = candidate_pair.map(lambda x:findSimilarity(x,b_c,businesses)).filter(lambda x:x[1]>=0.5).sortBy(lambda x:x[0][0]).map(lambda x:tuple(x[0])).collect()
with open(outputfile, "w") as f:
    f.write("business_id_1, business_id_2, similarity\n")
    for pair in similar_pair.collect():
        f.write(pair[0][0]+","+pair[0][1]+","+str(pair[1])+"\n")

# end = time.time()
# print(end-start)

# no false positive, can have a few false negative, so s must be high

# vali_data = sc.textFile("/Users/peiwendu/Downloads/public_data/pure_jaccard_similarity.csv")
# head = vali_data.first()
# vali_pair = vali_data.filter(lambda x: head not in x).map(lambda x:x.split(",")).map(lambda x:(x[0],x[1])).collect()
# ture_positive = len(set(similar_pair).intersection(set(vali_pair)))
# false_positive = len(set(similar_pair).difference(set(vali_pair)))
# print(false_positive)
# false_negaive = len(set(vali_pair).difference(set(similar_pair)))
# print(false_negaive)
# precision = ture_positive/(ture_positive+false_positive)
# recall = ture_positive/(ture_positive+false_negaive)
# print(precision)
# print(recall)

File: Locality-Sensitive-Hashing.scala
import org.apache.spark.{SparkConf, SparkContext}
import java.io.PrintWriter

object peiwen_du_task1 {
  def main(args: Array[String]): Unit = {
    val inputfile = args(0)
    val outputfile = args(1)

//    val inputfile = "/Users/peiwendu/Downloads/public_data/yelp_train.csv"
//    val outputfile = "peiwen_du_task1"

    val conf = new SparkConf()
    conf.setAppName("hw3.1")
    conf.setMaster("local[*]")
    val sc = new SparkContext(conf)
    sc.setLogLevel("WARN")

//    val start = System.currentTimeMillis()
    val rdd = sc.textFile(inputfile)
    val header = rdd.first()
    val test_data = rdd.filter(x => x != header).map(x => x.split(","))

    val users = test_data.map(_(0)).distinct().collect()
    val businesses = test_data.map(_(1)).distinct().collect()
    val user_index = scala.collection.mutable.HashMap.empty[String, Int]
    val business_index = scala.collection.mutable.HashMap.empty[String, Int]
    var index_of_users = 0
    var index_of_business = 0
    for (e <- users) {
      user_index(e) = index_of_users
      index_of_users += 1
    }
    for (e<- businesses){
      business_index(e) = index_of_business
      index_of_business+=1
    }

    val character_matrix = test_data.map(x=>(business_index(x(1)),user_index(x(0)))).groupByKey().mapValues(_.toList)
    val characters = character_matrix.collect()

    val b_c = scala.collection.mutable.HashMap.empty[Int, List[Int]]
    for (b<-characters){
      b_c(b._1) = b._2
//      println(b._2)
    }

    val m = users.length
//    println(m)

    val hashes = List(List(913, 51, 24593), List(2, 59, 53),List(232,45,201326611), List(54, 23, 769), List(622,321,6291469), List(1, 101, 193),List(431,21,805306457), List(17, 91, 1543),List(765,345,3145739),
    List(387, 552, 98317), List(11, 37, 3079),List(2123,222,100663319), List(2, 63, 97), List(543,342,1572869),List(145,234,25165843), List(41, 67, 6151),List(333,112,50331653),List(333,21,1610612741),
      List(291, 29, 12289), List(3, 79, 53),List(432,14,402653189), List(321,445,196613), List(951,340,393241),List(531,300,786433), List(473, 83, 49157),List(224,21,25165843), List(8, 119, 389),List(987,21,12582917))


    val signatiure = character_matrix.map(x=>{
      val tmp = scala.collection.mutable.ListBuffer.empty[Int]
      for (has <-hashes){
        var min = m+1
        for (u<-x._2){
          var hashvalue = ((has(0) *u +has(1))%has(2)) % m
          if (min >= hashvalue){
            min = hashvalue
          }
        }
        tmp += min
      }
      Tuple2(x._1,tmp.toList)
    })
//    for (i<-signatiure.take(2)){
//      println(i._1)
//      println(i._2)
//    }

//
//
    val candidate_pair = signatiure.flatMap(x=>{
      val bid = x._1
      val signature = x._2
      var bins = scala.collection.mutable.ListBuffer.empty[((Int, Tuple1[List[Int]]), List[Int])]
      for (i<- 0 until 28 by 2){
        bins += Tuple2(Tuple2(i/2, Tuple1(signature.slice(i, i+2))), List(bid))
      }
      bins.toList
    }).reduceByKey((x,y)=>List.concat(x, y)).filter(_._2.length >1).flatMap(x=> {
      val res = scala.collection.mutable.ListBuffer.empty[((Int, Int), Int)]
      val length = x._2.length
      val whole = x._2.sorted
//      println(whole)
      for (i <- 0 until length) {
        for (j <- i+1 until length) {
          res += Tuple2(Tuple2(whole(i), whole(j)), 1)
        }
      }
      res.toList
    }).reduceByKey((x, y) => x).map(x => x._1)

//        for (i<-candidate_pair.take(2)){
//          println(i._1)
//          println(i._2)
//        }

    val similar_pair = candidate_pair.map(x=> {
      val bid1 = x._1
      val bid2 = x._2
      val character1 = b_c(bid1).toSet
//      println(character1)
      val character2 = b_c(bid2).toSet
//      println(character2)
      val similarity = character1.intersect(character2).size.toFloat / character1.union(character2).size
//      println(similarity)
      val pair = List(businesses(bid1),businesses(bid2)).sorted
//      println(pair)
      val first = pair(0)
      val second = pair(1)
      Tuple3(first, second, similarity)
    }).filter(_._3>=0.5).sortBy(x => x._1)

//    for (i<-similar_pair.take(2)){
//      println(i._1)
//      println(i._2)
//      println(i._3)
//    }
    val pw = new PrintWriter(outputfile)
    pw.println("business_id_1, business_id_2, similarity")
    for (pair<- similar_pair.collect())
      pw.println(pair._1+","+pair._2+","+pair._3.toString)
    pw.close()


//    val end = System.currentTimeMillis()
//    println(end-start)
//
//    val similar_pairs = similar_pair.map(x=>(x._1,x._2)).collect()
//    val vali_data = sc.textFile("/Users/peiwendu/Downloads/public_data/pure_jaccard_similarity.csv")
//    val head = vali_data.first()
//    val vali_pair = vali_data.filter(x=>x != header).map(x=>x.split(",")).map(x=>(x(0),x(1))).collect()
//    val ture_positive = similar_pairs.toSet.intersect(vali_pair.toSet).size
//    val false_positive = similar_pairs.toSet.diff(vali_pair.toSet).size
//    println(false_positive)
//    val false_negaive = vali_pair.toSet.diff(similar_pairs.toSet).size
//    println(false_negaive)
//    val precision = ture_positive/(ture_positive+false_positive).toFloat
//    val recall = ture_positive/(ture_positive+false_negaive).toFloat
//
//    println(precision)
//    println(recall)

  }
}
